File: lenspr/parsers/ts_resolver.py
"""TypeScript cross-file resolution without Node.js.

Provides module resolution and export tracking for TypeScript/JavaScript projects.
Parses tsconfig.json for path aliases and tracks exports from parsed files.
"""

from __future__ import annotations

import json
import logging
from dataclasses import dataclass, field
from pathlib import Path
from typing import TYPE_CHECKING, Any

from lenspr.models import EdgeConfidence, Resolution

if TYPE_CHECKING:
    pass

logger = logging.getLogger(__name__)


@dataclass
class TsConfig:
    """Parsed tsconfig.json settings relevant for resolution."""

    base_url: str = "."
    paths: dict[str, list[str]] = field(default_factory=dict)
    root_dir: str = "."
    out_dir: str | None = None

    @classmethod
    def load(cls, project_root: Path) -> TsConfig:
        """Load tsconfig.json from project root."""
        config_path = project_root / "tsconfig.json"
        if not config_path.exists():
            # Try jsconfig.json for JS projects
            config_path = project_root / "jsconfig.json"

        if not config_path.exists():
            return cls()

        try:
            with open(config_path) as f:
                data = json.load(f)

            compiler_opts = data.get("compilerOptions", {})
            return cls(
                base_url=compiler_opts.get("baseUrl", "."),
                paths=compiler_opts.get("paths", {}),
                root_dir=compiler_opts.get("rootDir", "."),
                out_dir=compiler_opts.get("outDir"),
            )
        except Exception as e:
            logger.warning("Failed to parse tsconfig.json: %s", e)
            return cls()


@dataclass
class ExportInfo:
    """Information about an export from a file."""

    name: str
    node_id: str
    file_path: str
    is_default: bool = False
    is_type: bool = False  # export type { X }


class TypeScriptResolver:
    """
    Cross-file resolution for TypeScript/JavaScript projects.

    Resolution strategy:
    1. Parse tsconfig.json for path aliases
    2. Track exports from all parsed files
    3. Resolve imports by matching to exports
    4. Cache results for performance
    """

    def __init__(self, project_root: Path) -> None:
        self._project_root = project_root
        self._config = TsConfig.load(project_root)
        self._exports: dict[str, list[ExportInfo]] = {}  # file_path -> exports
        self._export_index: dict[str, ExportInfo] = {}  # "module.name" -> ExportInfo
        self._resolution_cache: dict[str, Resolution] = {}

        # Common external packages (don't try to resolve)
        self._external_packages = {
            "react",
            "react-dom",
            "next",
            "vue",
            "angular",
            "express",
            "lodash",
            "axios",
            "moment",
            "dayjs",
            "date-fns",
            "uuid",
            "zod",
            "yup",
        }

        logger.debug(
            "TypeScriptResolver initialized with baseUrl=%s, paths=%s",
            self._config.base_url,
            list(self._config.paths.keys()),
        )

    def register_exports(
        self, file_path: str, exports: list[dict[str, Any]]
    ) -> None:
        """Register exports from a parsed file.

        Args:
            file_path: Relative path from project root
            exports: List of export info dicts with keys:
                - name: exported name
                - node_id: full node ID in the graph
                - is_default: True if default export
                - is_type: True if type-only export
        """
        export_list = []
        for exp in exports:
            info = ExportInfo(
                name=exp["name"],
                node_id=exp["node_id"],
                file_path=file_path,
                is_default=exp.get("is_default", False),
                is_type=exp.get("is_type", False),
            )
            export_list.append(info)

            # Index by module.name for fast lookup
            module_id = self._file_to_module_id(file_path)
            key = f"{module_id}.{info.name}"
            self._export_index[key] = info

            # Also index default exports by module name only
            if info.is_default:
                self._export_index[module_id] = info

        self._exports[file_path] = export_list
        logger.debug("Registered %d exports from %s", len(export_list), file_path)

    def resolve(
        self,
        from_file: str,
        import_source: str,
        imported_name: str,
    ) -> Resolution:
        """Resolve an imported name to its definition.

        Args:
            from_file: File containing the import
            import_source: Import source (e.g., './utils', 'react', '@/lib/api')
            imported_name: Name being imported (e.g., 'useState', 'default')

        Returns:
            Resolution with node_id and confidence level
        """
        cache_key = f"{from_file}:{import_source}:{imported_name}"
        if cache_key in self._resolution_cache:
            return self._resolution_cache[cache_key]

        result = self._do_resolve(from_file, import_source, imported_name)
        self._resolution_cache[cache_key] = result
        return result

    def _do_resolve(
        self,
        from_file: str,
        import_source: str,
        imported_name: str,
    ) -> Resolution:
        """Internal resolution logic."""
        # Check if external package
        package_name = import_source.split("/")[0]
        if package_name in self._external_packages or not self._is_local_import(
            import_source
        ):
            return Resolution(
                node_id=f"{import_source}.{imported_name}",
                confidence=EdgeConfidence.EXTERNAL,
            )

        # Resolve module path
        resolved_path = self._resolve_module_path(import_source, from_file)
        if resolved_path is None:
            return Resolution(
                node_id=None,
                confidence=EdgeConfidence.UNRESOLVED,
                untracked_reason="module_not_found",
            )

        # Find export in resolved file
        module_id = self._file_to_module_id(resolved_path)

        # Try exact match first
        if imported_name == "default":
            key = module_id
        else:
            key = f"{module_id}.{imported_name}"

        if key in self._export_index:
            export = self._export_index[key]
            return Resolution(
                node_id=export.node_id,
                confidence=EdgeConfidence.RESOLVED,
            )

        # Try fuzzy match (name exists in file but not tracked as export)
        # This catches cases where we didn't parse exports correctly
        for exp in self._exports.get(resolved_path, []):
            if exp.name == imported_name:
                return Resolution(
                    node_id=exp.node_id,
                    confidence=EdgeConfidence.RESOLVED,
                )

        # Module exists but name not found - might be re-exported
        return Resolution(
            node_id=f"{module_id}.{imported_name}",
            confidence=EdgeConfidence.INFERRED,
            untracked_reason="export_not_found",
        )

    def _is_local_import(self, import_source: str) -> bool:
        """Check if import is local (not from node_modules)."""
        if import_source.startswith("."):
            return True
        if import_source.startswith("@/"):
            return True
        if import_source.startswith("~/"):
            return True
        # Check against tsconfig paths
        for pattern in self._config.paths:
            if import_source.startswith(pattern.rstrip("*")):
                return True
        return False

    def _resolve_module_path(
        self, import_source: str, from_file: str
    ) -> str | None:
        """Resolve import source to actual file path.

        Handles:
        - Relative imports: './utils', '../lib/helper'
        - Absolute imports via baseUrl: 'components/Button'
        - Path aliases: '@/lib/api', '~/utils'
        - Index files: './utils' -> './utils/index.ts'
        """
        from_dir = Path(from_file).parent

        # Handle path aliases first
        resolved = self._apply_path_aliases(import_source)
        if resolved != import_source:
            import_source = resolved

        # Handle relative imports
        if import_source.startswith("."):
            target = from_dir / import_source
        else:
            # Absolute import via baseUrl
            base = Path(self._config.base_url)
            target = base / import_source

        # Try to find the actual file
        return self._find_module_file(str(target))

    def _apply_path_aliases(self, import_source: str) -> str:
        """Apply tsconfig path aliases."""
        for pattern, replacements in self._config.paths.items():
            # Handle wildcard patterns like "@/*" -> ["src/*"]
            if pattern.endswith("*"):
                prefix = pattern[:-1]  # "@/"
                if import_source.startswith(prefix):
                    suffix = import_source[len(prefix) :]  # "lib/api"
                    if replacements:
                        replacement = replacements[0]
                        if replacement.endswith("*"):
                            return replacement[:-1] + suffix
                        return replacement + "/" + suffix
            elif import_source == pattern:
                if replacements:
                    return replacements[0]

        return import_source

    def _find_module_file(self, target: str) -> str | None:
        """Find actual file for a module path.

        Tries extensions and index files:
        - target.ts, target.tsx, target.js, target.jsx
        - target/index.ts, target/index.tsx, etc.

        Normalizes paths to resolve .. and . segments.
        """
        extensions = [".ts", ".tsx", ".js", ".jsx"]
        resolved_root = self._project_root.resolve()

        # Direct file match
        for ext in extensions:
            candidate = (self._project_root / f"{target}{ext}").resolve()
            try:
                if candidate.exists():
                    return str(candidate.relative_to(resolved_root))
            except ValueError:
                # Path is outside project root
                continue

        # Index file match
        for ext in extensions:
            candidate = (self._project_root / target / f"index{ext}").resolve()
            try:
                if candidate.exists():
                    return str(candidate.relative_to(resolved_root))
            except ValueError:
                continue

        # Already has extension
        target_path = Path(target)
        if target_path.suffix in extensions:
            candidate = (self._project_root / target).resolve()
            try:
                if candidate.exists():
                    return str(candidate.relative_to(resolved_root))
            except ValueError:
                pass

        return None

    def _file_to_module_id(self, file_path: str) -> str:
        """Convert file path to module-style ID."""
        path = Path(file_path)
        # Remove extension
        if path.suffix in (".ts", ".tsx", ".js", ".jsx"):
            path = path.with_suffix("")
        # Convert to dot notation
        parts = list(path.parts)
        # Handle index files
        if parts and parts[-1] == "index":
            parts = parts[:-1]
        return ".".join(parts) if parts else file_path

    def resolve_call(
        self,
        from_node: str,
        call_name: str,
        imports: dict[str, str],
    ) -> Resolution:
        """Resolve a function call to its definition.

        Args:
            from_node: Node ID of the caller (e.g., 'src.App.App')
            call_name: Name being called - can be:
                - Simple name: 'helper' (lookup in imports)
                - Import path: './utils/helpers.capitalize' (already resolved)
                - Method call: 'data.map'
            imports: Map of local_name -> import_source from the file

        Returns:
            Resolution with node_id and confidence level
        """
        # Extract from_file from from_node (convert dots to path)
        node_parts = from_node.split(".")
        if len(node_parts) > 1:
            # Remove the function name at the end (e.g., 'src.App.App' -> 'src/App')
            from_file = "/".join(node_parts[:-1])
        else:
            from_file = node_parts[0]

        # Check if call_name is already an import path (starts with './' or '../')
        if call_name.startswith("./") or call_name.startswith("../"):
            # Parse: './utils/helpers.capitalize' -> path='./utils/helpers', func='capitalize'
            last_segment = call_name.split("/")[-1]
            if "." in last_segment:
                # Last part has a dot, so it's path.function
                path_parts = call_name.rsplit(".", 1)
                import_source = path_parts[0]
                imported_name = path_parts[1] if len(path_parts) > 1 else "default"
            else:
                import_source = call_name
                imported_name = "default"

            # Resolve the import path to a module
            resolved_path = self._resolve_module_path(import_source, from_file + ".ts")
            if resolved_path:
                module_id = self._file_to_module_id(resolved_path)
                lookup_key = f"{module_id}.{imported_name}"

                if lookup_key in self._export_index:
                    return Resolution(
                        node_id=self._export_index[lookup_key].node_id,
                        confidence=EdgeConfidence.RESOLVED,
                    )

                # Check if it's in the exports list for that file
                for exp in self._exports.get(resolved_path, []):
                    if exp.name == imported_name:
                        return Resolution(
                            node_id=exp.node_id,
                            confidence=EdgeConfidence.RESOLVED,
                        )

            # Couldn't resolve fully
            return Resolution(
                node_id=call_name,
                confidence=EdgeConfidence.INFERRED,
                untracked_reason="import_path_not_resolved",
            )

        # Check if call is to an imported name (via imports dict)
        if call_name in imports:
            import_source = imports[call_name]
            return self.resolve(from_file, import_source, call_name)

        # Check if it's a method call like 'data.map' or 'react.useState'
        if "." in call_name:
            call_parts = call_name.split(".")
            first_part = call_parts[0]

            # Check if first part is an external package
            if first_part in self._external_packages:
                return Resolution(
                    node_id=call_name,
                    confidence=EdgeConfidence.EXTERNAL,
                )

            if first_part in imports:
                import_source = imports[first_part]
                return Resolution(
                    node_id=f"{import_source}.{call_name}",
                    confidence=EdgeConfidence.INFERRED,
                )

        # Check if defined in current module
        module_id = ".".join(from_node.split(".")[:-1]) if "." in from_node else from_node
        local_key = f"{module_id}.{call_name}"
        if local_key in self._export_index:
            return Resolution(
                node_id=self._export_index[local_key].node_id,
                confidence=EdgeConfidence.RESOLVED,
            )

        # Unresolved - could be built-in, external, or dynamic
        return Resolution(
            node_id=call_name,
            confidence=EdgeConfidence.INFERRED,
            untracked_reason="not_imported",
        )

    def clear_cache(self) -> None:
        """Clear resolution cache (call after file changes)."""
        self._resolution_cache.clear()

    def get_stats(self) -> dict[str, int]:
        """Get resolver statistics."""
        return {
            "tracked_files": len(self._exports),
            "total_exports": len(self._export_index),
            "cache_size": len(self._resolution_cache),
        }

File: lenspr/helpers/ts_resolver.js
#!/usr/bin/env node
/**
 * TypeScript Resolver using TypeScript Compiler API
 *
 * This script provides full type inference for resolving names in TypeScript/JavaScript.
 * It loads the project's tsconfig.json and uses the TypeScript type checker.
 *
 * Usage:
 *   Batch mode (stdin/stdout):
 *     echo '{"requests": [...]}' | node ts_resolver.js /path/to/project
 *
 *   Request format:
 *     {"requests": [
 *       {"id": "1", "file": "src/App.tsx", "line": 10, "column": 5},
 *       {"id": "2", "file": "src/utils.ts", "line": 20, "column": 10}
 *     ]}
 *
 *   Response format:
 *     {"results": [
 *       {"id": "1", "nodeId": "src/components/Button.Button", "confidence": "resolved"},
 *       {"id": "2", "nodeId": "react.useState", "confidence": "external"}
 *     ]}
 */

const ts = require('typescript');
const path = require('path');
const fs = require('fs');

// Confidence levels matching Python's EdgeConfidence
const Confidence = {
    RESOLVED: 'resolved',
    INFERRED: 'inferred',
    EXTERNAL: 'external',
    UNRESOLVED: 'unresolved'
};

// Known external packages
const EXTERNAL_PACKAGES = new Set([
    'react', 'react-dom', 'next', 'vue', 'angular', 'express',
    'lodash', 'axios', 'moment', 'dayjs', 'date-fns', 'uuid',
    'zod', 'yup', '@tanstack/react-query', 'swr', 'zustand',
    'redux', '@reduxjs/toolkit', 'mobx', 'immer', 'ramda'
]);

class TypeScriptResolver {
    constructor(projectRoot) {
        this.projectRoot = path.resolve(projectRoot);
        this.program = null;
        this.typeChecker = null;
        this.sourceFiles = new Map();

        this._initializeProgram();
    }

    _initializeProgram() {
        // Find all config files (tsconfig.json, jsconfig.json) in project
        const configPaths = this._findConfigFiles(this.projectRoot);

        let compilerOptions = {
            target: ts.ScriptTarget.ESNext,
            module: ts.ModuleKind.ESNext,
            moduleResolution: ts.ModuleResolutionKind.NodeJs,
            jsx: ts.JsxEmit.React,
            esModuleInterop: true,
            allowJs: true,
            checkJs: true,
            strict: false, // Don't need strict for resolution
            skipLibCheck: true,
            noEmit: true
        };

        let fileNames = [];

        // Process each config file found
        for (const configPath of configPaths) {
            const configFile = ts.readConfigFile(configPath, ts.sys.readFile);
            if (!configFile.error) {
                const parsed = ts.parseJsonConfigFileContent(
                    configFile.config,
                    ts.sys,
                    path.dirname(configPath)
                );
                // Merge options (later configs override)
                compilerOptions = { ...compilerOptions, ...parsed.options };
                // Collect all file names
                fileNames.push(...parsed.fileNames);
            }
        }

        // Deduplicate file names
        fileNames = [...new Set(fileNames)];

        // If no files from configs, find them manually
        if (fileNames.length === 0) {
            fileNames = this._findSourceFiles(this.projectRoot);
        }

        // Create the program
        this.program = ts.createProgram(fileNames, compilerOptions);
        this.typeChecker = this.program.getTypeChecker();

        // Cache source files
        for (const sourceFile of this.program.getSourceFiles()) {
            if (!sourceFile.isDeclarationFile) {
                const relativePath = path.relative(this.projectRoot, sourceFile.fileName);
                if (!relativePath.startsWith('node_modules')) {
                    this.sourceFiles.set(relativePath, sourceFile);
                    // Also store by absolute path
                    this.sourceFiles.set(sourceFile.fileName, sourceFile);
                }
            }
        }
    }

    _findConfigFiles(dir) {
        const configs = [];
        const configNames = ['tsconfig.json', 'jsconfig.json'];
        const skipDirs = new Set([
            'node_modules', '.git', 'dist', 'build', '.next',
            'coverage', '.cache', '__pycache__', '.lens'
        ]);

        const walk = (currentDir) => {
            try {
                const entries = fs.readdirSync(currentDir, { withFileTypes: true });
                for (const entry of entries) {
                    const fullPath = path.join(currentDir, entry.name);
                    if (entry.isDirectory()) {
                        if (!skipDirs.has(entry.name) && !entry.name.startsWith('.')) {
                            walk(fullPath);
                        }
                    } else if (entry.isFile() && configNames.includes(entry.name)) {
                        configs.push(fullPath);
                    }
                }
            } catch (e) {
                // Skip directories we can't read
            }
        };

        walk(dir);
        return configs;
    }

    _findSourceFiles(dir) {
        const files = [];
        const extensions = ['.ts', '.tsx', '.js', '.jsx'];
        const skipDirs = new Set([
            'node_modules', '.git', 'dist', 'build', '.next',
            'coverage', '.cache', '__pycache__'
        ]);

        const walk = (currentDir) => {
            try {
                const entries = fs.readdirSync(currentDir, { withFileTypes: true });
                for (const entry of entries) {
                    const fullPath = path.join(currentDir, entry.name);
                    if (entry.isDirectory()) {
                        if (!skipDirs.has(entry.name) && !entry.name.startsWith('.')) {
                            walk(fullPath);
                        }
                    } else if (entry.isFile()) {
                        const ext = path.extname(entry.name);
                        if (extensions.includes(ext)) {
                            files.push(fullPath);
                        }
                    }
                }
            } catch (e) {
                // Skip directories we can't read
            }
        };

        walk(dir);
        return files;
    }

    /**
     * Resolve a name at a specific position in a file.
     *
     * @param {string} filePath - Relative or absolute path to the file
     * @param {number} line - 1-based line number
     * @param {number} column - 0-based column number
     * @returns {Object} Resolution result with nodeId and confidence
     */
    resolve(filePath, line, column) {
        // Get the source file
        let sourceFile = this.sourceFiles.get(filePath);
        if (!sourceFile) {
            // Try absolute path
            const absolutePath = path.resolve(this.projectRoot, filePath);
            sourceFile = this.sourceFiles.get(absolutePath);
        }
        if (!sourceFile) {
            return {
                nodeId: null,
                confidence: Confidence.UNRESOLVED,
                reason: 'file_not_found'
            };
        }

        // Convert line/column to position
        const position = this._getPositionOfLineAndCharacter(sourceFile, line - 1, column);
        if (position < 0) {
            return {
                nodeId: null,
                confidence: Confidence.UNRESOLVED,
                reason: 'invalid_position'
            };
        }

        // Find the node at this position
        const node = this._findNodeAtPosition(sourceFile, position);
        if (!node) {
            return {
                nodeId: null,
                confidence: Confidence.UNRESOLVED,
                reason: 'no_node_at_position'
            };
        }

        // Try to resolve the node
        return this._resolveNode(node, sourceFile);
    }

    _getPositionOfLineAndCharacter(sourceFile, line, character) {
        try {
            return sourceFile.getPositionOfLineAndCharacter(line, character);
        } catch (e) {
            // Fall back to manual calculation
            const lines = sourceFile.text.split('\n');
            let position = 0;
            for (let i = 0; i < line && i < lines.length; i++) {
                position += lines[i].length + 1; // +1 for newline
            }
            return position + character;
        }
    }

    _findNodeAtPosition(sourceFile, position) {
        let result = null;

        const visit = (node) => {
            if (position >= node.getStart() && position < node.getEnd()) {
                result = node;
                ts.forEachChild(node, visit);
            }
        };

        visit(sourceFile);
        return result;
    }

    _resolveNode(node, sourceFile) {
        // Handle identifiers
        if (ts.isIdentifier(node)) {
            return this._resolveIdentifier(node, sourceFile);
        }

        // Handle property access (e.g., obj.method)
        if (ts.isPropertyAccessExpression(node)) {
            return this._resolvePropertyAccess(node, sourceFile);
        }

        // Handle call expressions
        if (ts.isCallExpression(node)) {
            return this._resolveNode(node.expression, sourceFile);
        }

        // Handle JSX elements
        if (ts.isJsxOpeningElement(node) || ts.isJsxSelfClosingElement(node)) {
            const tagName = node.tagName;
            if (ts.isIdentifier(tagName)) {
                return this._resolveIdentifier(tagName, sourceFile);
            }
            // Handle namespaced JSX like <Foo.Bar />
            if (ts.isPropertyAccessExpression(tagName)) {
                return this._resolvePropertyAccess(tagName, sourceFile);
            }
        }

        // Phase 3: Handle class heritage (extends clause)
        if (ts.isHeritageClause(node)) {
            // For `extends BaseClass`, resolve the base class
            if (node.token === ts.SyntaxKind.ExtendsKeyword && node.types && node.types.length > 0) {
                const extendsExpr = node.types[0].expression;
                return this._resolveNode(extendsExpr, sourceFile);
            }
        }

        // Handle expression with type arguments (BaseClass<T>)
        if (ts.isExpressionWithTypeArguments(node)) {
            return this._resolveNode(node.expression, sourceFile);
        }

        return {
            nodeId: null,
            confidence: Confidence.UNRESOLVED,
            reason: 'unsupported_node_type'
        };
    }

    _resolveIdentifier(node, sourceFile) {
        const name = node.text;

        // Get the symbol for this identifier
        const symbol = this.typeChecker.getSymbolAtLocation(node);
        if (!symbol) {
            return {
                nodeId: name,
                confidence: Confidence.INFERRED,
                reason: 'no_symbol'
            };
        }

        // Check if this is a default import
        const isDefault = this._isDefaultImport(node);

        // Follow aliases (imports) with improved re-export chain following
        const resolvedSymbol = this._followReexportChain(symbol);
        if (!resolvedSymbol) {
            return {
                nodeId: name,
                confidence: Confidence.INFERRED,
                reason: 'cannot_resolve_symbol'
            };
        }

        // Get the declaration
        const declarations = resolvedSymbol.getDeclarations();
        if (!declarations || declarations.length === 0) {
            return {
                nodeId: name,
                confidence: Confidence.INFERRED,
                reason: 'no_declarations'
            };
        }

        // For default imports, prefer the default export declaration
        let declaration = declarations[0];
        if (isDefault && declarations.length > 1) {
            const defaultDecl = declarations.find(d => this._isDefaultExportDeclaration(d));
            if (defaultDecl) {
                declaration = defaultDecl;
            }
        }

        const declarationFile = declaration.getSourceFile();

        // Check if it's from an external package
        const filePath = declarationFile.fileName;
        if (filePath.includes('node_modules')) {
            const packageName = this._extractPackageName(filePath);
            if (packageName) {
                return {
                    nodeId: `${packageName}.${name}`,
                    confidence: Confidence.EXTERNAL,
                    reason: 'external_package'
                };
            }
        }

        // Check if it's a type definition file
        if (declarationFile.isDeclarationFile) {
            return {
                nodeId: name,
                confidence: Confidence.EXTERNAL,
                reason: 'declaration_file'
            };
        }

        // Build the node ID
        const nodeId = this._buildNodeId(declaration, resolvedSymbol.name);
        return {
            nodeId,
            confidence: Confidence.RESOLVED,
            reason: 'resolved'
        };
    }

    /**
     * Check if this identifier is from a default import.
     * e.g., `import Button from './Button'`
     */
    _isDefaultImport(node) {
        let parent = node.parent;
        while (parent) {
            if (ts.isImportClause(parent)) {
                // This is `import X from '...'` where X is the default
                return parent.name === node;
            }
            parent = parent.parent;
        }
        return false;
    }

    /**
     * Check if a declaration is a default export.
     */
    _isDefaultExportDeclaration(declaration) {
        // Check for `export default`
        if (declaration.modifiers) {
            const hasExport = declaration.modifiers.some(m => m.kind === ts.SyntaxKind.ExportKeyword);
            const hasDefault = declaration.modifiers.some(m => m.kind === ts.SyntaxKind.DefaultKeyword);
            if (hasExport && hasDefault) {
                return true;
            }
        }

        // Check if it's an export assignment (export default X or export = X)
        if (ts.isExportAssignment(declaration)) {
            return !declaration.isExportEquals; // ES6 default, not CommonJS
        }

        return false;
    }

    _resolvePropertyAccess(node, sourceFile) {
        const propertyName = node.name.text;

        // Get the type of the expression being accessed
        const expressionType = this.typeChecker.getTypeAtLocation(node.expression);
        const symbol = expressionType.getSymbol();

        // Try to resolve the property via symbol lookup first
        const propSymbol = this.typeChecker.getSymbolAtLocation(node.name);
        if (propSymbol) {
            const resolved = this._followReexportChain(propSymbol);
            if (resolved) {
                const declarations = resolved.getDeclarations();
                if (declarations && declarations.length > 0) {
                    const declaration = declarations[0];
                    const declarationFile = declaration.getSourceFile();

                    // Check if external
                    if (declarationFile.fileName.includes('node_modules')) {
                        const packageName = this._extractPackageName(declarationFile.fileName);
                        return {
                            nodeId: `${packageName}.${propertyName}`,
                            confidence: Confidence.EXTERNAL
                        };
                    }

                    if (!declarationFile.isDeclarationFile) {
                        // Build node ID
                        const nodeId = this._buildNodeId(declaration, propertyName);
                        return {
                            nodeId,
                            confidence: Confidence.RESOLVED
                        };
                    }
                }
            }
        }

        // Phase 4: Try to resolve via apparent properties (type enumeration)
        if (expressionType) {
            try {
                const apparentProps = this.typeChecker.getApparentProperties(expressionType);
                const matchingProp = apparentProps.find(p => p.name === propertyName);

                if (matchingProp) {
                    const resolved = this._followReexportChain(matchingProp);
                    if (resolved) {
                        const declarations = resolved.getDeclarations();
                        if (declarations && declarations.length > 0) {
                            const declaration = declarations[0];
                            const declarationFile = declaration.getSourceFile();

                            // Check if external
                            if (declarationFile.fileName.includes('node_modules')) {
                                const packageName = this._extractPackageName(declarationFile.fileName);
                                return {
                                    nodeId: `${packageName}.${propertyName}`,
                                    confidence: Confidence.EXTERNAL
                                };
                            }

                            if (!declarationFile.isDeclarationFile) {
                                const nodeId = this._buildNodeId(declaration, propertyName);
                                return {
                                    nodeId,
                                    confidence: Confidence.RESOLVED
                                };
                            }
                        }
                    }
                }
            } catch (e) {
                // getApparentProperties can fail for some types
            }
        }

        // Fall back to type-based resolution
        if (symbol) {
            const symbolName = symbol.name;
            return {
                nodeId: `${symbolName}.${propertyName}`,
                confidence: Confidence.INFERRED,
                reason: 'type_based'
            };
        }

        return {
            nodeId: propertyName,
            confidence: Confidence.INFERRED,
            reason: 'cannot_resolve_property'
        };
    }

    /**
     * Follow re-export chains to find the actual declaration.
     * Handles: export { X } from './module', export * from './module', etc.
     * @param {ts.Symbol} symbol - The symbol to resolve
     * @param {number} maxDepth - Maximum chain depth to prevent infinite loops
     * @returns {ts.Symbol|null} The resolved symbol
     */
    _followReexportChain(symbol, maxDepth = 15) {
        let current = symbol;
        let depth = 0;
        const seen = new Set();

        while (current && depth < maxDepth) {
            // Use symbol name + declaration location as unique ID
            const symbolKey = this._getSymbolKey(current);
            if (seen.has(symbolKey)) {
                // Circular reference detected
                break;
            }
            seen.add(symbolKey);

            // Check if this is an alias (import or re-export)
            if (current.flags & ts.SymbolFlags.Alias) {
                try {
                    const aliased = this.typeChecker.getAliasedSymbol(current);
                    if (aliased && aliased !== current) {
                        current = aliased;
                        depth++;
                        continue;
                    }
                } catch (e) {
                    // getAliasedSymbol can throw for some edge cases
                    break;
                }
            }

            // Check for export value (module.exports = X)
            if (current.flags & ts.SymbolFlags.ExportValue) {
                try {
                    const exported = this.typeChecker.getExportSymbolOfSymbol?.(current);
                    if (exported && exported !== current) {
                        current = exported;
                        depth++;
                        continue;
                    }
                } catch (e) {
                    break;
                }
            }

            // No more chain to follow
            break;
        }

        return current;
    }

    /**
     * Get a unique key for a symbol to detect circular references.
     */
    _getSymbolKey(symbol) {
        const declarations = symbol.getDeclarations?.();
        if (declarations && declarations.length > 0) {
            const decl = declarations[0];
            const sf = decl.getSourceFile();
            return `${symbol.name}@${sf.fileName}:${decl.getStart()}`;
        }
        return `${symbol.name}@unknown`;
    }

    /**
     * Legacy method for backwards compatibility.
     * @deprecated Use _followReexportChain instead
     */
    _resolveSymbol(symbol) {
        return this._followReexportChain(symbol);
    }

    _extractPackageName(filePath) {
        const nodeModulesIndex = filePath.lastIndexOf('node_modules');
        if (nodeModulesIndex === -1) return null;

        const afterNodeModules = filePath.substring(nodeModulesIndex + 'node_modules/'.length);
        const parts = afterNodeModules.split('/');

        // Handle scoped packages (@org/package)
        if (parts[0].startsWith('@') && parts.length > 1) {
            return `${parts[0]}/${parts[1]}`;
        }
        return parts[0];
    }

    _buildNodeId(declaration, name) {
        const sourceFile = declaration.getSourceFile();
        const filePath = sourceFile.fileName;

        // Get relative path from project root
        let relativePath = path.relative(this.projectRoot, filePath);

        // Remove extension
        relativePath = relativePath.replace(/\.(tsx?|jsx?)$/, '');

        // Convert path separators to dots
        let moduleId = relativePath.replace(/[/\\]/g, '.');

        // Handle index files
        if (moduleId.endsWith('.index')) {
            moduleId = moduleId.slice(0, -6);
        }

        // Find the containing function/class/module
        const container = this._findContainer(declaration);
        if (container && container !== name) {
            return `${moduleId}.${container}.${name}`;
        }

        return `${moduleId}.${name}`;
    }

    _findContainer(node) {
        let current = node.parent;
        while (current) {
            if (ts.isFunctionDeclaration(current) && current.name) {
                return current.name.text;
            }
            if (ts.isClassDeclaration(current) && current.name) {
                return current.name.text;
            }
            if (ts.isMethodDeclaration(current) && current.name) {
                if (ts.isIdentifier(current.name)) {
                    // Get class name too
                    const classNode = current.parent?.parent;
                    if (ts.isClassDeclaration(classNode) && classNode.name) {
                        return `${classNode.name.text}.${current.name.text}`;
                    }
                    return current.name.text;
                }
            }
            current = current.parent;
        }
        return null;
    }

    /**
     * Resolve a call at a specific location.
     * This is a higher-level method that handles common patterns.
     */
    resolveCall(filePath, line, column, callName) {
        // First try position-based resolution
        const result = this.resolve(filePath, line, column);
        if (result.confidence === Confidence.RESOLVED ||
            result.confidence === Confidence.EXTERNAL) {
            return result;
        }

        // Check if it's a known external package call
        if (callName && callName.includes('.')) {
            const parts = callName.split('.');
            const packageName = parts[0].toLowerCase();
            if (EXTERNAL_PACKAGES.has(packageName)) {
                return {
                    nodeId: callName,
                    confidence: Confidence.EXTERNAL,
                    reason: 'known_external_package'
                };
            }
        }

        return result;
    }

    /**
     * Process multiple requests in batch.
     */
    processBatch(requests) {
        const results = [];
        for (const req of requests) {
            const result = this.resolve(req.file, req.line, req.column);
            results.push({
                id: req.id,
                ...result
            });
        }
        return results;
    }

    /**
     * Get resolver statistics.
     */
    getStats() {
        return {
            projectRoot: this.projectRoot,
            sourceFileCount: this.sourceFiles.size,
            hasTypeChecker: this.typeChecker !== null
        };
    }
}

// Main entry point
function main() {
    const args = process.argv.slice(2);

    if (args.length === 0) {
        console.error('Usage: node ts_resolver.js <project_root>');
        console.error('       Reads JSON requests from stdin, writes results to stdout');
        process.exit(1);
    }

    const projectRoot = args[0];

    // Check if project exists
    if (!fs.existsSync(projectRoot)) {
        console.error(JSON.stringify({
            error: `Project root does not exist: ${projectRoot}`
        }));
        process.exit(1);
    }

    // Initialize resolver
    let resolver;
    try {
        resolver = new TypeScriptResolver(projectRoot);
    } catch (e) {
        console.error(JSON.stringify({
            error: `Failed to initialize resolver: ${e.message}`
        }));
        process.exit(1);
    }

    // Read input from stdin
    let input = '';
    process.stdin.setEncoding('utf8');

    process.stdin.on('data', (chunk) => {
        input += chunk;
    });

    process.stdin.on('end', () => {
        try {
            const data = JSON.parse(input);

            if (data.command === 'stats') {
                console.log(JSON.stringify(resolver.getStats()));
                return;
            }

            if (!data.requests || !Array.isArray(data.requests)) {
                console.log(JSON.stringify({
                    error: 'Invalid input: expected {"requests": [...]}'
                }));
                return;
            }

            const results = resolver.processBatch(data.requests);
            console.log(JSON.stringify({ results }));

        } catch (e) {
            console.log(JSON.stringify({
                error: `Failed to process requests: ${e.message}`
            }));
        }
    });
}

// Export for testing
module.exports = { TypeScriptResolver, Confidence };

// Run if called directly
if (require.main === module) {
    main();
}
